libcaf_core/caf/exec_main.hpp

```diff
@@ -105,8 +105,28 @@ int exec_main(F fun, int argc, char** argv) {
 
 } // namespace caf
 
+namespace caf::detail {
+
+template <class... Module>
+auto do_init_host_system(type_list<Module...>, type_list<>) {
+  return std::make_tuple(Module::init_host_system()...);
+}
+
+template <class... Module, class T, class... Ts>
+auto do_init_host_system(type_list<Module...>, type_list<T, Ts...>) {
+  if constexpr (detail::has_init_host_system_v<T>) {
+    return do_init_host_system(type_list<Module..., T>{}, type_list<Ts...>{});
+  } else {
+    return do_init_host_system(type_list<Module...>{}, type_list<Ts...>{});
+  }
+}
```

Comment:
This dispatches the `init_host_system()` call to either `Module` or `T`?

Reply:
This entire thing is basically filtering the `type_list<...>`. Each `T` with `has_init_host_system_v<T> == true` ends up in the `Module...` list. Once the entire type list is consumed, the remainder is just `type_list<>` and the first overload kicks in to call `init_host_system()` on all remaining types. Or it just returns `std::make_tuple()` if nothing matched.